pkg/resourcetracker/app.go

```diff
@@ -100,17 +113,62 @@ func CreateComponentRevisionResourceTracker(ctx context.Context, cli client.Clie
 	return createResourceTracker(ctx, cli, app, getComponentRevisionResourceTrackerName(app), v1beta1.ResourceTrackerTypeComponentRevision)
 }
 
-// ListApplicationResourceTrackers list resource trackers for application with all historyRTs sorted by version number
-func ListApplicationResourceTrackers(ctx context.Context, cli client.Client, app *v1beta1.Application) (rootRT *v1beta1.ResourceTracker, currentRT *v1beta1.ResourceTracker, historyRTs []*v1beta1.ResourceTracker, crRT *v1beta1.ResourceTracker, err error) {
-	metrics.ListResourceTrackerCounter.WithLabelValues("application").Inc()
+func newResourceTrackerFromApplicationResourceTracker(appRt *unstructured.Unstructured) (*v1beta1.ResourceTracker, error) {
+	rt := &v1beta1.ResourceTracker{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(appRt.Object, rt); err != nil {
+		return nil, err
+	}
+	namespace := metav1.NamespaceDefault
+	if ns := appRt.GetNamespace(); ns != "" {
+		namespace = ns
+	}
+	rt.SetName(appRt.GetName() + "-" + namespace)
+	rt.SetNamespace("")
+	rt.SetGroupVersionKind(v1beta1.ResourceTrackerKindVersionKind)
+	return rt, nil
+}
+
+func listApplicationResourceTrackers(ctx context.Context, cli client.Client, app *v1beta1.Application) ([]v1beta1.ResourceTracker, error) {
 	rts := v1beta1.ResourceTrackerList{}
-	if err = cli.List(ctx, &rts, client.MatchingLabels{
+	err := cli.List(ctx, &rts, client.MatchingLabels{
 		oam.LabelAppName:      app.Name,
 		oam.LabelAppNamespace: app.Namespace,
-	}); err != nil {
+	})
+	if err == nil {
+		return rts.Items, nil
+	}
+	if !kerrors.IsForbidden(err) && !kerrors.IsUnauthorized(err) {
+		return nil, err
+	}
+	appRts := &unstructured.UnstructuredList{}
+	appRts.SetGroupVersionKind(applicationResourceTrackerGroupVersionKind)
+	if err = cli.List(ctx, appRts, client.MatchingLabels{
+		oam.LabelAppName: app.Name,
+	}, client.InNamespace(app.Namespace)); err != nil {
+		if velaerrors.IsCRDNotExists(err) {
+			return nil, errors.New("no permission for ResourceTracker and apiextensions are not enabled")
+		}
+		return nil, err
```

Comment:
can we check if the prism not installed/availabile and nofity user to check that?

Reply:
@wonderflow Actually, L148 is checking for that.

Reply:
it's only check CRD exists, does it mean the server is running?

Reply:
If the server is not running, the APIService will not work properly, then it will return CRD not exists. The underlying logic is okay.

Reply:
ok